### PR TITLE
Auto-Sklearn Integration

### PR DESCRIPTION
**Pull request by MinuraPunchihewa**
This is a draft PR for the Auto-Sklearn ML integration.

### CODE REVIEW

**Comment by MinuraPunchihewa** [comment]
Hey @paxcema,
I finally got this integration to work.. I think. I've still got a ways to go though, I'm only supporting classification here.

By the way, I am sorry to trouble you with multiple PRs.

Could you please take a look at what I have done?

Given below are the steps I have followed to get this running using a Google Sheets data source,
```
CREATE ML_ENGINE autosklearn_engine FROM autosklearn;

CREATE MODEL mindsdb.auto_sklearn_predictor
FROM sheets_datasource_4
    (SELECT * FROM winequality_red) 
PREDICT quality AS quality
USING
engine='autosklearn_engine';

SELECT *
FROM mindsdb.models
WHERE name='auto_sklearn_predictor';

SELECT m.quality
FROM sheets_datasource_5.winequality_red_predict AS t
JOIN mindsdb.auto_sklearn_predictor AS m;
```

I noticed that when making predictions, there is a column called `__mindsdb_row_id` being added. I had to remove this to get my `predict()` function to work. I did not see this in the other ML handlers though. Do you know what's happening there?

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]
Please first solve the flake issues https://github.com/mindsdb/mindsdb/actions/runs/4566631743/jobs/8059406028?pr=5312#step:7:80

**Comment by MinuraPunchihewa** [comment]
Hey @ZoranPandovski, it does not look like these issues are coming up from the integration?
image

Is this maybe because I have not merged the most up-to-date code?

**Comment by paxcema** [comment]
Also, `__mindsdb_row_id` indeed is used for some internal things (in particular with the lightwood integration). It should be deprecated soon, but in the meantime it's all right to just `pop()` it inside your integration to get it working fine 👍

File: mindsdb/integrations/handlers/autosklearn_handler/requirements.txt
Comment (paxcema) [inline mindsdb/integrations/handlers/autosklearn_handler/requirements.txt:2]: This package seems like it's not being used yet?

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:30]: Just to double check: I take it that `df.drop` does not do this in-place, correct? Otherwise this would not work.

**Review by paxcema — COMMENTED** [review]
In general this is looking very good @MinuraPunchihewa!

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:25]: Oh, you mean the `time_left_for_this_task` attribute, yes? That sounds like a great idea. There are actually quite a few parameters that go into the classifier as shown here,
https://automl.github.io/auto-sklearn/master/api.html

Do you think I should allow users control over all of these? Or shall I just pick out some of the more important ones for now?

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:30]: No, it does not. Only if the `inplace` argument is set to `True` would it be executed that way.

File: mindsdb/integrations/handlers/autosklearn_handler/requirements.txt
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/requirements.txt:2]: Are you talking about `mindsdb-evaluator`? The reason I have added this as a dependency, Patricio, is because the previous versions of this package used an old version of `sklearn`, which is not compatible with `auto-sklearn`. You might remember that I opened a PR on `mindsdb-evaluator` to resolve this?

Comment (paxcema) [reply mindsdb/integrations/handlers/autosklearn_handler/requirements.txt:2]: Aah yes, I do remember now, that makes sense!

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (paxcema) [reply mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:25]: Based on experience, let's go for the most important ones and we can enable other more specific options if there is interest from users.

Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:25]: Sounds good, Patricio. Thanks.

**Comment by ZoranPandovski** [comment]
> Hey @ZoranPandovski, it does not look like these issues are coming up from the integration? image
> 
> Is this maybe because I have not merged the most up-to-date code?

Now it pass successfully, thanks. @paxcema This is ready to merge

**Review by ZoranPandovski — APPROVED** [review]

**Comment by MinuraPunchihewa** [comment]
Hey @ZoranPandovski and @paxcema,
Please give me a little more time. I wanted to add a few more tweaks.

**Comment by MinuraPunchihewa** [comment]
Hey @paxcema,
I believe this is ready for your review now.

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:23]: Perhaps out of scope for this PR (I'll let you decide), but it would be pretty cool if we used `type_infer` here to automatically figure out if it's a regression or classification problem (or something else that is not supported!).

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:32]: I think we need an else here, just to raise an Exception when the task is not supported.

File: mindsdb/integrations/handlers/autosklearn_handler/config.py
Comment (paxcema) [inline mindsdb/integrations/handlers/autosklearn_handler/config.py:8]: Seeing as both configs are, right now, identical, I wonder if it's best to define a `BaseConfig` where all shared params are defined so that they can be potentially overridden in the two existing classes, plus adding any task-specific params in them?

It's a nitpick, so feel free to ignore this one.

**Review by paxcema — CHANGES_REQUESTED** [review]
We can merge this after fixing the check for unsupported task types. The other two comments are nice to haves, but not critical.

Awesome job @MinuraPunchihewa! We may have started with issues, but this is a lovely end result, congratulations 🥳

File: mindsdb/integrations/handlers/autosklearn_handler/config.py
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/config.py]: No, this is an awesome suggestion. It's done!

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py:35]: Done!

File: mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/autosklearn_handler/autosklearn_handler.py]: Love this. It's done!

**Comment by MinuraPunchihewa** [comment]
Hey @paxcema,
Thank you very much. This is probably one of my favorite contributions yet. I've been struggling quite a while with this.

**Comment by MinuraPunchihewa** [comment]
Apologies, I close the PR by mistake.

**Pull request merged**
